hal/hal_remote/source/hal_socket_transmitter.cpp

```diff
@@ -0,0 +1,163 @@
+    bool print_port) {
+  if (!setup_connection_done) {
+    hal_socket_transmitter::error_code res = setup_connection(true);
+    if (res != 0) {
+      return res;
+    }
+    setup_connection_done = true;
+  }
+  if (listen() != 0) {
+    last_error = hal_socket_transmitter::listen_failed;
+    return last_error;
+  }
+  if (print_port) {
+    printf("Listening on port %d\n", get_port());
+    fflush(stdout);
+  }
+  if (accept() == -1) {
+    last_error = hal_socket_transmitter::accept_failed;
+    return last_error;
+  }
+  last_error = hal_socket_transmitter::success;
+  return last_error;
+}
+
+/// @brief make a connection to a server
+/// @return true if successful, otherwise will set last_error to errno
+hal_socket_transmitter::error_code hal_socket_transmitter::make_connection() {
+  if (!setup_connection_done) {
+    last_error = setup_connection(false);
+    if (last_error != hal_socket_transmitter::success) {
+      return last_error;
+    }
+    setup_connection_done = true;
+  }
+  last_error = connect();
+  return last_error;
+}
+
+bool hal_socket_transmitter::receive(void *data, uint32_t size) {
+  uint32_t data_to_read = size;
+  uint32_t offset = 0;
+
+  // repeatedly recv until `size` bytes is read.
+  do {
+    int res = recv(fd_to_use, ((char *)data) + offset, data_to_read, 0);
+    // If recv returns 0, this indicates the connection has been dropped
+    // It's not an error as such but we are not able to continue
+    if (res == 0) {
+      is_connected = false;
+      fd_to_use = 0;
+      last_error = hal_socket_transmitter::connection_closed;
+      return false;
+    }
+    if (res == -1) {
+      last_error = hal_socket_transmitter::recv_error;
+      return false;
+    }
+    data_to_read = data_to_read - res;
+    offset += res;
+  } while (data_to_read);
+  last_error = hal_socket_transmitter::success;
+  return true;
+}
+
+/// @brief Send `size` bytes of `data` with an optional flush
+bool hal_socket_transmitter::send(const void *data, uint32_t size, bool flush) {
+  const int ret = ::send(fd_to_use, data, size, 0);
+  if (ret == -1) {
+    last_error = hal_socket_transmitter::send_error;
+    return false;
+  }
+  last_error = hal_socket_transmitter::success;
+  return true;
+}
+
+hal_socket_transmitter::error_code hal_socket_transmitter::connect() {
+  int res = ::connect(sock, (struct sockaddr *)&server_address,
+                      sizeof(server_address));
+  if (res != -1) {
+    fd_to_use = sock;
+    is_connected = true;
+    return hal_socket_transmitter::success;
+  } else {
+    return hal_socket_transmitter::connect_failed;
+  }
+}
+
+hal_socket_transmitter::error_code hal_socket_transmitter::setup_connection(
+    bool server) {
+  // @return 0 if success, otherwise -1 and errno set
+  sock = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+  server_address.sin_family = AF_INET;
+  server_address.sin_port = htons(port_requested);
+  server_address.sin_addr.s_addr = INADDR_ANY;
```

Comment:
Always binding to all interfaces, with no way to override, seems like a possible security concern.

Reply:
what would you prefer here? I'm fairly new to sockets.

Reply:
See e.g. Apache's configuration, https://httpd.apache.org/docs/2.4/bind.html, where users can specify the IP address to listen to in addition to the port. How this should look in code I'm not entirely sure of, it depends on how this PR later ends up used, which I have not looked at yet.

Reply:
I'm happy to extend this to an optional address and default to this (and document that) - would you be happy with that?

Reply:
That seems okay, though keep in mind that depending on the address, we may also need `AF_INET6` instead of `AF_INET`.

Question: since the purpose of the HAL process is to take untrusted code from a remote client and execute it, what is the overall security going to look like?

Reply:
it does add risk, but the user should be aware that this does remotely run processes, so your suggestion of limiting it to a particular IP address sounds good. Also it's not got any superuser priviledges.
For that matter, you can override the libhal as a shared library and grab things that way.